[RFC] Health: add clipboard check

mhinz opened the pull request:
I think that `:h clipboard` is short and concise already, so it's the only suggestion made when no supported clipboard tool was found.

mhinz committed:
provider/clipboard.vim: refactor

mhinz committed:
health/provider.vim: add clipboard check

bfredl commented:
:+1: Maybe also introduce `provider#clipboard#Error` in analogy with `provider#python#Error` ? (so that `has('clipboard')` can be invoked without showing the message)

phanimahesh commented:
In check_tmux, when checking escape time, the condition checks for > 500 but message reads greater than 300. Is that intentional and due to something weird, or a mistake? It should be corrected if it's a mistake and a comment should be added if it's something weird.

justinmk commented:
@phanimahesh The check should be changed to check 300.

mhinz commented on runtime/autoload/health/nvim.vim:
```diff
@@ -90,6 +90,20 @@ function! s:check_performance() abort
   endif
 endfunction
 
+function! s:check_clipboard() abort
```
Done.

mhinz committed:
provider/clipboard.vim: never show a warning on sourcing

Never throw an error when provider/clipboard.vim is sourced for the first time.
Save the error instead and expose it via `provider#clipboard#Error()`, mimicking
provider/python.vim.

mhinz commented:
I also added a third commit that suppresses an error when the clipboard provider is sourced for the first time and no clipboard tool was found.

The error gets exposed via `provider#clipboard#Error()`.

justinmk commented on runtime/autoload/provider/clipboard.vim:
```diff
@@ -31,34 +31,51 @@ function! s:try_cmd(cmd, ...)
+
+function! provider#clipboard#Error() abort
+  return s:err
+endfunction
```
will this close https://github.com/neovim/neovim/issues/3097 ?

mhinz replied:
Yea, the only way to see that error would be by calling `provider#clipboard#Error()`, which does not happen at the moment.

justinmk replied:
is that blocking this? maybe leave it for another PR?

bfredl replied:
I plan to do that, and I don't think it is blocking this PR.

bfredl commented:
Hmm, apparently `provider#xx#Error()` is not ever used. `:python x` just shows "Unknown function: provider#python#Call" instead of the actual error (if I set `g:python_host_prog = "blargh"`). I thought `eval_call_provider` would call this if `eval_has_provider` was false. Maybe it did at some point and then we regressed.

mhinz commented:
> Hmm, apparently provider#xx#Error() is not ever used.

So it wasn't me being crazy after all. ;-)

But I guess we can keep that commit anyway. Nobody wants to get distracted by warning messages at start and if something isn't working, `:CheckHealth` should be the first thing to check anyway.

Maybe we shouuld use `provider#...#Error()` as a convention in the first place.

bfredl commented:
No, apparently #2208 introduced `#Error()` but no calls to it...

> But I guess we can keep that commit anyway. 

:+1: 
For full diagnostics the user should use `:CheckHealth` but it still would be better if `:python` and `"+y` shows a relevant one or two-line message instead of totally useless "Unknown function: provider#python#Call". I will look into that.

mhinz commented:
If @bfredl looks into issuing that warning message by calling `provider#clipboard#Error()`, this can be merged as-is.

justinmk commented on runtime/autoload/health/provider.vim:
```diff
@@ -45,6 +45,19 @@ function! s:download(url) abort
+  if empty(clipboard_tool)
+    call health#report_warn(
+          \ "No clipboard tool found. Using the system clipboard won't work.",
+          \ ['See ":h clipboard"'])
```
should spell out `:help`.

Future idea: highlight help tags in `:CheckHealth` and make `K` work on them.
